### PR TITLE
Fix glitch filter to be streamed

### PR DESCRIPTION
**Pull request by falkn**
Fix glitch filter to be streamed to better handle short and many glitches.

Also
* Fix a bug in handling of too short pulse stream
* Make sure pin is input in setup.
* Add comments to IRQ.

### CODE REVIEW

File: Rx433.cpp [hunk 1]
```diff
@@ -62,63 +62,58 @@ bool Handle(std::vector<Pulse> buf) {
...
+  }
+  last_changed = now;
+
+  // Queue is full, drop pulse
+  if (pulse_stream_queue.size() > 10) return;
+
+  // Create pulse
   // Need to invert as p represents state before transition.
-  Pulse p = {now, now-last_changed, digitalRead(rxPin_)};
+  Pulse p = {now, delta_us, state};
```
Comment (henrikssn) [inline Rx433.cpp]: Doesn't it make more sense to move this up and get rid of now / state / delta_us variables?

File: Rx433.cpp [hunk 2]
```diff
@@ -62,63 +62,58 @@ bool Handle(std::vector<Pulse> buf) {
...
+  // Ignore any pulse (1 or 0) shorter than kGlitchUs.
+  if (!pulse_stream.empty() && delta_us < kGlitchUs) {
+    last_changed -= pulse_stream.back().delta_us;
+    pulse_stream.pop_back();
+    return;
+  }
+  last_changed = now;
+
+  // Queue is full, drop pulse
+  if (pulse_stream_queue.size() > 10) return;
```
Comment (henrikssn) [inline Rx433.cpp:85]: I would move this up (to save CPU). It should be first thing after static variables

File: Rx433.cpp [hunk 3]
```diff
@@ -62,63 +62,58 @@ bool Handle(std::vector<Pulse> buf) {
...
+      pulse_stream_queue.push_back(std::move(pulse_stream));
+    }
     pulse_stream_done = false;
-    // Need to reset as it was move:d above.
+    // Need to reset also when it was move:d above.
     pulse_stream.clear();
     if (IsSync(p)) {
       sync_pulse_us = p.delta_us;
-      pulse_stream.push_back(p);
+      pulse_stream.push_back(std::move(p));
```
Comment (henrikssn) [inline Rx433.cpp]: Nit: you don't gain anything with this move as p is stack allocated

File: Rx433.cpp [hunk 4]
```diff
@@ -62,63 +62,58 @@ bool Handle(std::vector<Pulse> buf) {
...
+  int delta_us = now - last_changed;
+
+  // Ignore very short glitches that we missed.
+  if (state == last_state) return;
+  last_state = state;
+
+  // Streamed glich filter
+  // Ignore any pulse (1 or 0) shorter than kGlitchUs.
+  if (!pulse_stream.empty() && delta_us < kGlitchUs) {
+    last_changed -= pulse_stream.back().delta_us;
```
Comment (henrikssn) [inline Rx433.cpp:78]: Another slightly more elegant way to do this:
```
pulse_stream.pop_back();
last_changed = pulse_stream.back().time_us;
```

**Review by henrikssn — CHANGES_REQUESTED** [review]
Overall: It would be interesting to run a benchmark to understand how this affects runtime of the ISR. Not blocking on that though as I don't see any particularly expensive operations in there.

File: Rx433.cpp (hunk 2 above)
Comment (falkn) [reply Rx433.cpp:85]: I moved it down as I think the other checks (same state, glitch) will trigger more and thus save more CPU. These ought to be more time critical as well.

File: Rx433.cpp (hunk 4 above)
Comment (falkn) [reply Rx433.cpp:78]: Done

File: Rx433.cpp (hunk 1 above)
Comment (falkn) [reply Rx433.cpp]: Done

File: Rx433.cpp (hunk 3 above)
Comment (falkn) [reply Rx433.cpp]: Done

File: Rx433.cpp [hunk 5]
```diff
@@ -65,17 +65,17 @@ void ICACHE_RAM_ATTR rxISR() {
...
+  if (p.state == last_state) return;
+  last_state = p.state;
 
   // Streamed glich filter
   // Ignore any pulse (1 or 0) shorter than kGlitchUs.
-  if (!pulse_stream.empty() && delta_us < kGlitchUs) {
-    last_changed -= pulse_stream.back().delta_us;
+  if (!pulse_stream.empty() && p.delta_us < kGlitchUs) {
+    last_changed = pulse_stream.back().time_us;
     pulse_stream.pop_back();
```
Comment (henrikssn) [inline Rx433.cpp:79]: These lines needs to be reversed, as last_changed should contain the time_us of the last element in the stream after the last one was dropped.

**Review by henrikssn — APPROVED** [review]
Just one thing, feel free to merge after fixing that.

File: Rx433.cpp (hunk 5 above)
Comment (falkn) [reply Rx433.cpp:79]: That will unfortunately not work when there are exactly 1 pulse in the list already. Reverting to use delta_ms instead to also support that case.

Actually, I think time_us is never really used so could be removed fully from the Pulse struct.

**Pull request merged**